File: scripts/build-declarations.py
#!/usr/bin/env python

import ast
import os
import json
import re
import sys
from collections import OrderedDict

TOOL_ENTRY_FUNC = "run"


def main(is_tool=True):
    scriptfile = sys.argv[1]
    is_tool = os.path.dirname(scriptfile) == "tools"

    with open(scriptfile, "r", encoding="utf-8") as f:
        contents = f.read()

    functions = extract_functions(contents, is_tool)
    declarations = []
    for function in functions:
        func_name, docstring, func_args = function
        description, params = parse_docstring(docstring)
        declarations.append(
            build_declaration(func_name, description, params, func_args)
        )

    if is_tool:
        name = os.path.splitext(os.path.basename(scriptfile))[0]
        if declarations:
            declarations = declarations[0:1]
            declarations[0]["name"] = name

    print(json.dumps(declarations, indent=2))


def extract_functions(contents: str, is_tool: bool):
    tree = ast.parse(contents)
    output = []
    for node in ast.walk(tree):
        if not isinstance(node, ast.FunctionDef):
            continue
        func_name = node.name
        if func_name.startswith("_"):
            continue
        if is_tool and func_name != TOOL_ENTRY_FUNC:
            continue
        docstring = ast.get_docstring(node) or ""
        func_args = OrderedDict()
        for arg in node.args.args:
            arg_name = arg.arg
            arg_type = get_arg_type(arg.annotation)
            func_args[arg_name] = arg_type
        output.append((func_name, docstring, func_args))
    return output


def get_arg_type(annotation) -> str:
    if annotation is None:
        return ""
    elif isinstance(annotation, ast.Name):
        return annotation.id
    elif isinstance(annotation, ast.Subscript):
        if isinstance(annotation.value, ast.Name):
            type_name = annotation.value.id
            if type_name == "List":
                child = get_arg_type(annotation.slice)
                return f"list[{child}]"
            if type_name == "Literal":
                literals = [ast.unparse(el) for el in annotation.slice.elts]
                return f"{'|'.join(literals)}"
            if type_name == "Optional":
                child = get_arg_type(annotation.slice)
                return f"{child}?"
    return "any"


def parse_docstring(docstring: str):
    lines = docstring.splitlines()
    description = ""
    rawParams = []
    is_in_args = False
    for line in lines:
        if not is_in_args:
            if line.startswith("Args:"):
                is_in_args = True
            else:
                description += f"\n{line}"
            continue
        else:
            if re.search(r"^\s+", line):
                rawParams.append(line.strip())
            else:
                break
    params = {}
    for rawParam in rawParams:
        name, type_, param_description = parse_param(rawParam)
        params[name] = (type_, param_description)
    return (description.strip(), params)


def parse_param(raw_param: str):
    name = ""
    description = ""
    type_from_comment = ""
    if ":" in raw_param:
        name, description = raw_param.split(":", 1)
        name = name.strip()
        description = description.strip()
    else:
        name = raw_param
    if " " in name:
        name, type_from_comment = name.split(" ", 1)
        type_from_comment = type_from_comment.strip()

    if type_from_comment.startswith("(") and type_from_comment.endswith(")"):
        type_from_comment = type_from_comment[1:-1]
    type_parts = [value.strip() for value in type_from_comment.split(",")]
    type_ = type_parts[0]
    if "optional" in type_parts[1:]:
        type_ = f"{type_}?"

    return (name, type_, description)


def build_declaration(
    name: str, description: str, params: dict, args: OrderedDict[str, str]
) -> dict[str, dict]:
    declaration = {
        "name": name,
        "description": description,
        "parameters": {
            "type": "object",
            "properties": {},
        },
    }
    schema = declaration["parameters"]
    required_params = []
    for arg_name, arg_type in args.items():
        type_ = arg_type
        description = ""
        required = True
        if params.get(arg_name):
            param_type, description = params[arg_name]
            if not type_:
                type_ = param_type
        if type_.endswith("?"):
            type_ = type_[:-1]
            required = False
        try:
            property = build_property(type_, description)
        except:
            raise ValueError(f"Unable to parse arg '{arg_name}' of function '{name}'")
        schema["properties"][arg_name] = property
        if required:
            required_params.append(arg_name)
    if required_params:
        schema["required"] = required_params
    return declaration


def build_property(type_: str, description: str):
    property = {}
    if "|" in type_:
        property["type"] = "string"
        property["enum"] = type_.replace("'", "").split("|")
    elif type_ == "bool":
        property["type"] = "boolean"
    elif type_ == "str":
        property["type"] = "string"
    elif type_ == "int":
        property["type"] = "integer"
    elif type_ == "float":
        property["type"] = "number"
    elif type_ == "list[str]":
        property["type"] = "array"
        property["items"] = {"type": "string"}
    elif type_ == "":
        property["type"] = "string"
    else:
        raise ValueError(f"Unsupported type `{type_}`")
    property["description"] = description
    return property


if __name__ == "__main__":
    main()

File: scripts/build-declarations.js
#!/usr/bin/env node

const fs = require("fs");
const path = require("path");

const TOOL_ENTRY_FUNC = "run";

function main() {
  const scriptfile = process.argv[2];
  const isTool = path.dirname(scriptfile) == "tools";
  const contents = fs.readFileSync(process.argv[2], "utf8");
  const functions = extractFunctions(contents, isTool);
  let declarations = [];
  for (const { funcName, jsdoc } of functions) {
    const { description, params } = parseJsDoc(jsdoc, funcName);
    if (!description) continue;
    const declaration = buildDeclaration(funcName, description, params);
    declarations.push(declaration);
  }
  if (isTool) {
    const name = getBasename(scriptfile);
    if (declarations.length > 0) {
      declarations = declarations.slice(0, 1);
      declarations[0].name = name;
    }
  }
  console.log(JSON.stringify(declarations, null, 2));
}

/**
 * @param {string} contents
 * @param {bool} isTool
 */
function extractFunctions(contents, isTool) {
  const output = [];
  const lines = contents.split("\n");
  let isInComment = false;
  let jsdoc = "";
  let incompleteComment = "";
  for (let line of lines) {
    if (/^\s*\/\*/.test(line)) {
      isInComment = true;
      incompleteComment += `\n${line}`;
    } else if (/^\s*\*\//.test(line)) {
      isInComment = false;
      incompleteComment += `\n${line}`;
      jsdoc = incompleteComment;
      incompleteComment = "";
    } else if (isInComment) {
      incompleteComment += `\n${line}`;
    } else {
      if (!jsdoc || line.trim() === "") {
        continue;
      }
      if (isTool) {
        if (new RegExp(`^export (async )?function ${TOOL_ENTRY_FUNC}|^exports\.${TOOL_ENTRY_FUNC}`).test(line)) {
          output.push({
            funcName: TOOL_ENTRY_FUNC,
            jsdoc,
          });
        }
      } else {
        let match = /^export (async )?function ([A-Za-z0-9_]+)/.exec(line);
        let funcName = null;
        if (match) {
          funcName = match[2];
        }
        if (!funcName) {
          match = /^exports\.([A-Za-z0-9_]+) = (async )?function /.exec(line);
          if (match) {
            funcName = match[1];
          }
        }
        if (funcName && !funcName.startsWith("_")) {
          output.push({ funcName, jsdoc });
        }
      }
      jsdoc = "";
    }
  }
  return output;
}

/**
 * @param {string} jsdoc
 * @param {string} funcName,
 */
function parseJsDoc(jsdoc, funcName) {
  const lines = jsdoc.split("\n");
  let description = "";
  const rawParams = [];
  let tag = "";
  for (let line of lines) {
    line = line.replace(/^\s*(\/\*\*|\*\/|\*)/, "").trim();
    let match = /^@(\w+)/.exec(line);
    if (match) {
      tag = match[1];
    }
    if (!tag) {
      description += `\n${line}`;
    } else if (tag == "property") {
      if (match) {
        rawParams.push(line.slice(tag.length + 1).trim());
      } else {
        rawParams[rawParams.length - 1] += `\n${line}`;
      }
    }
  }
  const params = [];
  for (const rawParam of rawParams) {
    try {
      params.push(parseParam(rawParam));
    } catch (err) {
      throw new Error(
        `Unable to parse function '${funcName}' of jsdoc '@property ${rawParam}'`,
      );
    }
  }
  return {
    description: description.trim(),
    params,
  };
}

/**
 * @typedef {ReturnType<parseParam>} Param
 */

/**
 * @param {string} rawParam
 */
function parseParam(rawParam) {
  const regex = /^{([^}]+)} +(\S+)( *- +| +)?/;
  const match = regex.exec(rawParam);
  if (!match) {
    throw new Error(`Invalid jsdoc comment`);
  }
  const type = match[1];
  let name = match[2];
  const description = rawParam.replace(regex, "");

  let required = true;
  if (/^\[.*\]$/.test(name)) {
    name = name.slice(1, -1);
    required = false;
  }
  let property = buildProperty(type, description);
  return { name, property, required };
}

/**
 * @param {string} type
 * @param {string} description
 */
function buildProperty(type, description) {
  type = type.toLowerCase();
  const property = {};
  if (type.includes("|")) {
    property.type = "string";
    property.enum = type.replace(/'/g, "").split("|");
  } else if (type === "boolean") {
    property.type = "boolean";
  } else if (type === "string") {
    property.type = "string";
  } else if (type === "integer") {
    property.type = "integer";
  } else if (type === "number") {
    property.type = "number";
  } else if (type === "string[]") {
    property.type = "array";
    property.items = { type: "string" };
  } else {
    throw new Error(`Unsupported type '${type}'`);
  }
  property.description = description;
  return property;
}

/**
 * @param {string} name
 * @param {string} description
 * @param {Param[]} params
 */
function buildDeclaration(name, description, params) {
  const declaration = {
    name,
    description,
    parameters: {
      type: "object",
      properties: {},
    },
  };
  const schema = declaration.parameters;
  const requiredParams = [];
  for (const { name, property, required } of params) {
    schema.properties[name] = property;
    if (required) {
      requiredParams.push(name);
    }
  }
  if (requiredParams.length > 0) {
    schema.required = requiredParams;
  }
  return declaration;
}

/**
 * @param {string} filePath
 */
function getBasename(filePath) {
  const filenameWithExt = filePath.split(/[/\\]/).pop();

  const lastDotIndex = filenameWithExt.lastIndexOf(".");

  if (lastDotIndex === -1) {
    return filenameWithExt;
  }

  return filenameWithExt.substring(0, lastDotIndex);
}

main();
